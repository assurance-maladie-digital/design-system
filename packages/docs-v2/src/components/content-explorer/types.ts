export interface TreeviewItem {
    label: string;
    path: string;
    icon?: string;
    children?: TreeviewItem[]
}