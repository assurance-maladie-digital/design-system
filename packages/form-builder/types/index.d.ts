import { PluginFunction } from 'vue';

// Types for the library usage
// eg. Vue.use(FormBuilder, options)

/* eslint-disable @typescript-eslint/no-empty-interface */
export interface FormBuilderOptions {}

export interface FormBuilderPlugin {
	install: PluginFunction<FormBuilderOptions>;
}

declare const FormBuilder: FormBuilderPlugin;
export default FormBuilder;
