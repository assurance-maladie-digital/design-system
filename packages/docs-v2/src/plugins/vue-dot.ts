import Vue from 'vue';

// Import dayjs locale
// import 'dayjs/locale/fr';

// Import VueDot from src
import VueDot from '@cnamts/vue-dot/src/index';

// Import the theme
import { icons } from '~/assets/icons';

import InputFacade from 'vue-input-facade';
Vue.use(InputFacade);

Vue.use(VueDot, {
	theme: {
		icons
	}
});
