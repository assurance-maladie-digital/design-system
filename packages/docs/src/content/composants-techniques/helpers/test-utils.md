---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
