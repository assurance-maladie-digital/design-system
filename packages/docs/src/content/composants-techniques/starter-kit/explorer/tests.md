Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
