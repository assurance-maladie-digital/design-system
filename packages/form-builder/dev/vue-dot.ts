import Vue from 'vue';

import 'dayjs/locale/fr';

import InputFacade from 'vue-input-facade';
Vue.use(InputFacade);

import VueDot from '@cnamts/vue-dot';

Vue.use(VueDot);
