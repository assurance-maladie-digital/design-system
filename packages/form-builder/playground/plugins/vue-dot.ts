import Vue from 'vue';

// Import dayjs locale
import 'dayjs/locale/fr';

// Import VueDot
import VueDot from '@cnamts/vue-dot/';

// Import the theme styles
import '../theme/theme.scss';

import VueTheMask from 'vue-the-mask';
Vue.use(VueTheMask);

export const LOCAL_STORAGE_CONTROL = {
	version: 1
};

Vue.use(VueDot, {
	localStorageControl: LOCAL_STORAGE_CONTROL
});
