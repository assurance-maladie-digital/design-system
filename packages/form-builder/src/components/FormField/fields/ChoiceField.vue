<template>
	<div>
		<component
			:is="choiceField"
			:value="choiceValue.value"
			:items="field.items"
			:options="fieldOptions"
			:multiple="field.multiple"
			:error-messages="errorMessages.value"
			@change="choiceUpdated"
		/>

		<template v-if="otherField">
			<VExpandTransition
				v-if="choiceSelected"
				hide-on-leave
			>
				<div
					v-if="showOtherTextareaField"
					class="vd-other-field mt-4"
				>
					<h4
						v-if="otherField.label"
						class="text-body-1 mb-1"
					>
						{{ otherField.label }}
					</h4>

					<VTextarea
						ref="otherFieldRef"
						v-bind="otherField.fieldOptions"
						:value="otherFieldValue"
						:disabled="!otherActive"
						:rows="1"
						:error-messages="errorMessages.other"
						auto-grow
						outlined
						class="vd-form-input"
						@change="otherUpdated"
					/>
				</div>
			</VExpandTransition>

			<VTextField
				v-else
				v-bind="otherField.fieldOptions"
				:value="otherFieldValue"
				:background-color="otherFieldValue ? 'accent' : undefined"
				:dark="Boolean(otherFieldValue)"
				:error-messages="errorMessages.other"
				color="accent"
				dense
				outlined
				class="vd-form-input"
				@input="setOtherValue"
				@change="otherUpdated"
			/>
		</template>
	</div>
</template>

<script lang="ts">
	import Component, { mixins } from 'vue-class-component';

	import { Refs } from '@cnamts/vue-dot/src/types';

	import { FieldComponent } from '../mixins/fieldComponent';

	import { IFieldMap } from '../mixins/fieldMap';

	import { ChoiceValue, OtherFieldValue, ChoiceFieldValue, OtherField, ChoiceFieldErrorMessages } from '../types';

	const MixinsDeclaration = mixins(FieldComponent);

	// We import them all because the form can use any of them
	import ChoiceAutocompleteField from './ChoiceAutocompleteField.vue';
	import ChoiceSelectField from './ChoiceSelectField.vue';
	import ChoiceButtonField from './ChoiceButtonField.vue';
	import ChoiceSliderField from './ChoiceSliderField.vue';
	import TextareaField from './TextareaField.vue';

	@Component<ChoiceField>({
		components: {
			ChoiceAutocompleteField,
			ChoiceSelectField,
			ChoiceButtonField,
			ChoiceSliderField,
			TextareaField
		},
		watch: {
			'field.value': {
				handler(value: ChoiceValue) {
					if (!value) {
						this.choiceValue = {
							value: null
						};
					} else {
						this.choiceValue = value;
					}

					/**
					 * Set the other local value if the other value is not null
					 * to keep the local value up for the user
					 */
					if (this.choiceValue?.other) {
						this.otherFieldValue = this.choiceValue.other;
					}
				},
				immediate: true,
				deep: true
			}
		}
	})
	export default class ChoiceField extends MixinsDeclaration {
		$refs!: Refs<{
			otherFieldRef: HTMLInputElement;
		}>;

		choiceValue = {} as ChoiceValue;

		otherFieldValue: OtherFieldValue = null;

		selectFieldMap: IFieldMap = {
			select: 'ChoiceSelectField',
			choiceAutocomplete: 'ChoiceAutocompleteField',
			choiceButton: 'ChoiceButtonField',
			choiceSlider: 'ChoiceSliderField'
		};

		get otherField(): OtherField | undefined {
			return this.field.other;
		}

		get showOtherTextareaField(): boolean {
			const otherSelectedChoice = this.otherField?.selectedChoice;

			if (!this.choiceSelected && !otherSelectedChoice) {
				return false;
			}

			const choiceFieldValue = this.choiceValue.value;

			if (choiceFieldValue === otherSelectedChoice) {
				return true;
			}

			if (!Array.isArray(choiceFieldValue) || !choiceFieldValue.length) {
				return false;
			}

			if (choiceFieldValue.includes(otherSelectedChoice)) {
				return true;
			}

			return false;
		}

		get choiceSelected(): boolean {
			return this.otherField?.selectedChoice !== undefined && this.otherField?.selectedChoice !== null;
		}

		get otherActive(): boolean {
			const choiceFieldValue = this.choiceValue?.value;
			const selectedChoice = this.field.other?.selectedChoice;

			if (!this.choiceSelected) {
				return false;
			} else if (Array.isArray(choiceFieldValue)) {
				return choiceFieldValue.includes(selectedChoice);
			} else {
				return choiceFieldValue === selectedChoice;
			}
		}

		get choiceField(): string {
			const metadataType = this.field.fieldOptions?.type as string || undefined;

			return metadataType ? this.selectFieldMap[metadataType] : this.selectFieldMap.select;
		}

		get errorMessages(): ChoiceFieldErrorMessages {
			const errorMessages = this.fieldOptions?.errorMessages;

			if (typeof errorMessages === 'string') {
				return {
					value: [errorMessages],
					other: undefined
				};
			}

			if (Array.isArray(errorMessages)) {
				return {
					value: errorMessages,
					other: undefined
				};
			}

			if (typeof errorMessages === 'object') {
				let { value, other } = errorMessages as ChoiceFieldErrorMessages;

				value = typeof value === 'string' ? [value] : value;
				other = typeof other === 'string' ? [other] : other;

				return {
					value,
					other
				};
			}

			return {};
		}

		choiceUpdated(choiceFieldValue: ChoiceFieldValue): void {
			this.choiceValue.value = choiceFieldValue;

			if (this.choiceValue.value && !this.otherField?.selectedChoice) {
				this.otherFieldValue = null;
			}

			if (this.otherActive) {
				this.$nextTick(() => this.$refs.otherFieldRef.focus());

				this.choiceValue.other = this.otherFieldValue;
			} else {
				this.choiceValue.other = null;
			}

			this.emitChangeEvent(this.choiceValue);
		}

		setOtherValue(otherFieldValue: OtherFieldValue): void {
			this.otherFieldValue = otherFieldValue?.length ? otherFieldValue : null;

			if (this.otherFieldValue && !this.field.multiple) {
				this.choiceValue.value = null;
			}
		}

		otherUpdated(otherFieldValue: OtherFieldValue): void {
			this.otherFieldValue = otherFieldValue?.length ? otherFieldValue : null;
			this.choiceValue.other = this.otherFieldValue;

			this.emitChangeEvent(this.choiceValue);
		}
	}
</script>
