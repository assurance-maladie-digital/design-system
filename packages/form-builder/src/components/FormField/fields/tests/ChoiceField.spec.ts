import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@cnamts/vue-dot/tests';
import { Field } from './../../types';

import ChoiceField from '../ChoiceField.vue';

let wrapper: Wrapper<Vue>;

const testField: Field = {
	type: 'select',
	title: 'Votre taille',
	items: [
		{
			text: 'S',
			value: 0
		},
		{
			text: 'M',
			value: 1
		},
		{
			text: 'Autre',
			value: 'autre'
		}
	],
	fieldOptions: {
		label: 'Classic field',
		outlined: true
	}
};

describe('ChoiceField', () => {
	it('renders correctly', () => {
		wrapper = mountComponent(ChoiceField, {
			propsData: {
				field: {
					...testField,
					value: {
						value: '0'
					}
				}
			}
		});

		expect(wrapper).toMatchSnapshot();
	});

	it('renders the other field', () => {
		wrapper = mountComponent(ChoiceField, {
			propsData: {
				field: {
					...testField,
					value: {
						value: null,
						other: 'XXL'
					},
					other: {
						label: 'Autre taille',
						fieldOptions: {
							hint: 'Veuillez entrez votre taille'
						}
					}
				}
			}
		});

		expect(wrapper).toMatchSnapshot();
	});

	it('renders the other field when the corresponding choice is selected', () => {
		wrapper = mountComponent(ChoiceField, {
			propsData: {
				field: {
					...testField,
					value: {
						value: 'autre',
						other: 'XXL'
					},
					other: {
						label: 'Autre taille',
						selectedChoice: 'autre',
						fieldOptions: {
							hint: 'Veuillez entrez votre taille'
						}
					}
				}
			}
		});

		expect(wrapper).toMatchSnapshot();
	});

	it('renders with single error message', () => {
		wrapper = mountComponent(
			ChoiceField,
			{
				propsData: {
					field: {
						...testField,
						fieldOptions: {
							type: 'choiceButton',
							errorMessages: 'Erreur'
						}
					}
				}
			},
			true
		);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders with multiple error messages', () => {
		wrapper = mountComponent(
			ChoiceField,
			{
				propsData: {
					field: {
						...testField,
						fieldOptions: {
							type: 'choiceButton',
							errorMessages: [
								'Erreur 1',
								'Erreur 2'
							]
						}
					}
				}
			},
			true
		);

		expect(wrapper).toMatchSnapshot();
	});
});
