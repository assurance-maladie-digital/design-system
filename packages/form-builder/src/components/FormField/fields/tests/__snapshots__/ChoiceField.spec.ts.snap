// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChoiceField renders correctly 1`] = `
<div>
  <choiceselectfield-stub items="[object Object],[object Object],[object Object]" value="0" options="[object Object]"></choiceselectfield-stub>
  <!---->
</div>
`;

exports[`ChoiceField renders the other field 1`] = `
<div>
  <choiceselectfield-stub items="[object Object],[object Object],[object Object]" options="[object Object]"></choiceselectfield-stub>
  <vtextfield-stub color="accent" dark="true" errorcount="1" errormessages="" messages="" rules="" successmessages="" value="XXL" backgroundcolor="accent" dense="true" hint="Veuillez entrez votre taille" loaderheight="2" clearicon="$clear" outlined="true" type="text" class="vd-form-input"></vtextfield-stub>
</div>
`;

exports[`ChoiceField renders the other field when the corresponding choice is selected 1`] = `
<div>
  <choiceselectfield-stub items="[object Object],[object Object],[object Object]" value="autre" options="[object Object]"></choiceselectfield-stub>
  <vexpandtransition-stub mode="in-out" hide-on-leave="">
    <div class="vd-other-field mt-4">
      <h4 class="text-body-1 mb-1">
        Autre taille
      </h4>
      <vtextarea-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="XXL" backgroundcolor="" hint="Veuillez entrez votre taille" loaderheight="2" clearicon="$clear" outlined="true" type="text" autogrow="true" rowheight="24" rows="1" class="vd-form-input"></vtextarea-stub>
    </div>
  </vexpandtransition-stub>
</div>
`;

exports[`ChoiceField renders with multiple error messages 1`] = `
<div>
  <div class="vd-select-btn-field vd-form-input" type="choiceButton">
    <div class="vd-select-btn-field-toggle d-flex overflow-y-auto primary--text v-item-group theme--light v-btn-toggle flex-column" role="listbox"><button type="button" value="0" class="v-btn v-btn--outlined theme--light elevation-0 v-size--default primary--text" style="height: auto; min-height: 40px;" role="option"><span class="v-btn__content"><span class="text-body-1">
				S
			</span>
        <div class="spacer"></div> <span aria-hidden="true" class="v-icon notranslate theme--light white--text flex-shrink-0 ml-1" style="visibility: hidden;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></span>
      </button><button type="button" value="1" class="v-btn v-btn--outlined theme--light elevation-0 v-size--default primary--text" style="height: auto; min-height: 40px;" role="option"><span class="v-btn__content"><span class="text-body-1">
				M
			</span>
        <div class="spacer"></div> <span aria-hidden="true" class="v-icon notranslate theme--light white--text flex-shrink-0 ml-1" style="visibility: hidden;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></span>
      </button><button type="button" value="autre" class="v-btn v-btn--outlined theme--light elevation-0 v-size--default primary--text" style="height: auto; min-height: 40px;" role="option"><span class="v-btn__content"><span class="text-body-1">
				Autre
			</span>
        <div class="spacer"></div> <span aria-hidden="true" class="v-icon notranslate theme--light white--text flex-shrink-0 ml-1" style="visibility: hidden;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></span>
      </button></div>
    <p class="v-messages error--text px-3 mt-2 mb-0 theme--light">
      Erreur 1
    </p>
    <p class="v-messages error--text px-3 mt-2 mb-0 theme--light">
      Erreur 2
    </p>
  </div>
  <!---->
</div>
`;

exports[`ChoiceField renders with single error message 1`] = `
<div>
  <div class="vd-select-btn-field vd-form-input" type="choiceButton">
    <div class="vd-select-btn-field-toggle d-flex overflow-y-auto primary--text v-item-group theme--light v-btn-toggle flex-column" role="listbox"><button type="button" value="0" class="v-btn v-btn--outlined theme--light elevation-0 v-size--default primary--text" style="height: auto; min-height: 40px;" role="option"><span class="v-btn__content"><span class="text-body-1">
				S
			</span>
        <div class="spacer"></div> <span aria-hidden="true" class="v-icon notranslate theme--light white--text flex-shrink-0 ml-1" style="visibility: hidden;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></span>
      </button><button type="button" value="1" class="v-btn v-btn--outlined theme--light elevation-0 v-size--default primary--text" style="height: auto; min-height: 40px;" role="option"><span class="v-btn__content"><span class="text-body-1">
				M
			</span>
        <div class="spacer"></div> <span aria-hidden="true" class="v-icon notranslate theme--light white--text flex-shrink-0 ml-1" style="visibility: hidden;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></span>
      </button><button type="button" value="autre" class="v-btn v-btn--outlined theme--light elevation-0 v-size--default primary--text" style="height: auto; min-height: 40px;" role="option"><span class="v-btn__content"><span class="text-body-1">
				Autre
			</span>
        <div class="spacer"></div> <span aria-hidden="true" class="v-icon notranslate theme--light white--text flex-shrink-0 ml-1" style="visibility: hidden;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></span>
      </button></div>
    <p class="v-messages error--text px-3 mt-2 mb-0 theme--light">
      Erreur
    </p>
  </div>
  <!---->
</div>
`;
